ROLLO Website (#53)
arfc/rollo: docs/source/usersguide/checkpoint_rollo.rst

.. _checkpoint_rollo:

.. image:: ../pics/rollo-logo.png
  :width: 450
  :alt: rollo-logo

======================
ROLLO Checkpoint File
======================

The checkpoint file is a `serialized pickle file 
<https://docs.python.org/3/library/pickle.html>`_ and is based on the `DEAP 
checkpointing system <https://deap.readthedocs.io/en/master/tutorials/advanced/checkpoint.html?highlight=rndstate#checkpointing>`_. 

The checkpoint file holds the **ROLLO** simulation results and acts as a restart 
file. Thus, if a **ROLLO** simulation ends prematurely, users can use the checkpoint 
file to restart the code from the most recent population and continue the simulation.

Restarting Simulation
=====================
The checkpoint file updates after each generation runs successfully. 
If the **ROLLO** simulation fails half way through a generation, the user can
restart the simulation from the previous generation. 

> Therefore, if you're running a ROLLO simulation with 10 generations and it fails 

Reviewer: ```suggestion
Therefore, if you're running a **ROLLO** simulation with 10 generations and it fails 
```

> during generation 3, you can restart the ROLLO simulation from generation 2. 

Reviewer: ```suggestion
during generation 3, you can restart the **ROLLO** simulation from generation 2. 
```

   * - ``all`` [``ind_naming``]
     - dict, key: input parameter name, value: index, corresponds to each individual's input parameter index in ``all`` [``populations``] 
   * - ``all`` [``oup_naming``]
     - dict, key: output parameter name, value: index, corresponds to each individual's output parameter index in ``all`` [``outputs``] 
   * - ``all`` [``populations``]
     - list of populations, each population list contains a list of reactor models individuals
   * - ``all`` [``outputs``]
     - list of population outputs, each population list contains a list of reactor model individual's output parameters

Each reactor model individual is an Ind class type with attributes and is a simple 
list (`DEAP Individual class <https://deap.readthedocs.io/en/master/tutorials/basic/part1.html#individual>`_).
The following code snippet demonstrates a reactor model individual's attributes. 

.. image:: ../pics/ind-demonstrate.png
  :width: 600
  :alt: Notebook snippet

Descriptions of the reactor model individual's attributes: 

.. list-table::
   :widths: 25 75
   :header-rows: 1

   * - Attribute 
     - Description
   * - ``fitness``

>      - fitness tuple holds the objective values. The sign refers to whether the objective is maximize or minimize. 

Reviewer: ```suggestion
     - fitness tuple holds the objective values. The sign refers to whether the objective is to maximize or minimize. 
```

   * - ``gen``
     - generation 
   * - ``num``
     - reactor model index in generation
   * - ``output`` 

>      - tuple of reactor model individual's output parameters

Reviewer: ```suggestion
     - tuple of reactor model individual output parameters
```

.. _checkpoint_rollo:

.. image:: ../pics/rollo-logo.png
  :width: 450
  :alt: rollo-logo

======================
ROLLO Checkpoint File
======================

The checkpoint file is a `serialized pickle file 
<https://docs.python.org/3/library/pickle.html>`_ and is based on the `DEAP 
checkpointing system <https://deap.readthedocs.io/en/master/tutorials/advanced/checkpoint.html?highlight=rndstate#checkpointing>`_. 

The checkpoint file holds the **ROLLO** simulation results and acts as a restart 
file. Thus, if a **ROLLO** simulation ends prematurely, users can use the checkpoint 

> file to restart the code from the most recent population and continue the simulation.

Reviewer 2: ```suggestion
file.
```

You can delete this since it's repeated immediately after

Restarting Simulation
=====================
The checkpoint file updates after each generation runs successfully. 

> If the **ROLLO** simulation fails half way through a generation, the user can

Reviewer 2: ```suggestion
If the **ROLLO** simulation fails during a generation, the user can
```

   * - ``population`` 
     - the final population (list of reactor model individuals)
   * - ``generation``
     - number of generations (int)
   * - ``hall of fame``
     - deap.tools.HallOfFame object (`see deap documentation <https://deap.readthedocs.io/en/master/api/tools.html#deap.tools.HallOfFame>`_)
   * - ``logbook``
     - deap.tools.Logbook object (`see deap documentation <https://deap.readthedocs.io/en/master/api/tools.html#logbook>`_)
   * - ``rndstate`` 
     -  simulation's random state (used when restarting simulation)
   * - ``all`` [``ind_naming``]
     - dict, key: input parameter name, value: index, corresponds to each individual's input parameter index in ``all`` [``populations``] 
   * - ``all`` [``oup_naming``]
     - dict, key: output parameter name, value: index, corresponds to each individual's output parameter index in ``all`` [``outputs``] 
   * - ``all`` [``populations``]
     - list of populations, each population list contains a list of reactor models individuals
   * - ``all`` [``outputs``]
     - list of population outputs, each population list contains a list of reactor model individual's output parameters

> Each reactor model individual is an Ind class type with attributes and is a simple 

Reviewer 2: ```suggestion
Each reactor model individual is an :class:`Ind` class type with attributes and is a simple 
```
